x/upgrade/abci.go

```diff
@@ -53,49 +53,83 @@ func BeginBlocker(k keeper.Keeper, ctx sdk.Context, _ abci.RequestBeginBlock) {
 		},
 	)
 
-	// To make sure clear upgrade is executed at the same block
+	// If the plan's block height has passed, then it must be the executed version
+	// All major and minor releases are REQUIRED to execute on the scheduled block height
 	if plan.ShouldExecute(ctx) {
 		// If skip upgrade has been set for current height, we clear the upgrade plan
 		if k.IsSkipHeight(ctx.BlockHeight()) {
-			skipUpgradeMsg := fmt.Sprintf("UPGRADE \"%s\" SKIPPED at %d: %s", plan.Name, plan.Height, plan.Info)
-			ctx.Logger().Info(skipUpgradeMsg)
-
-			// Clear the upgrade plan at current height
-			k.ClearUpgradePlan(ctx)
+			skipUpgrade(k, ctx, plan)
 			return
 		}
+		// If we don't have an upgrade handler for this upgrade name, then we need to shutdown
+		if !k.HasHandler(plan.Name) {
+			panicUpgradeNeeded(k, ctx, plan)
+		}
+		applyUpgrade(k, ctx, plan)
+		return
+	}
 
+	// If running a pending minor release, apply the upgrade if handler is present
+	// Minor releases are allowed to run before the scheduled upgrade height, but not required to.
+	if plan.UpgradeDetails().IsMinorRelease() {
+		// if the pending upgrade is skipped, don't apply it
+		if k.IsSkipHeight(plan.Height) {
+			skipUpgrade(k, ctx, plan)
```

Comment:
how will this behave if the upgrade is applied 100 blocks early, but then the skip height (height of upgrade) arrives, and then the `skipUpgrade` logic runs, but the upgrade has already been applied? will this bork the binary?

Reply:
wait actually, question, would this cause issue with apphash because upgrading early will modify the stored state in upgrade keeper?